feat(backends/s3): add warmup support for check command

gilbsgilbs opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

Follow-up from https://github.com/restic/restic/pull/5173, implements warmup support from cold storage in `check` command.

See also https://github.com/restic/restic/issues/3202

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

No, but it follows closely what has been done in https://github.com/restic/restic/pull/5173 .

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
  - ➡️ ~~Testing this change requires mocking the `Repository` struct in `checker`. It looks like a lot of work to test a mock… WDYT? Is there a better way?~~
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
  - ➡️ ~~I've edited the existing changelog because it is not released yet, but if there's a release before this PR gets merged, then I'll create a new one.~~
- [x] I'm done! This pull request is ready for review.

gilbsgilbs committed:
feat(backends/s3): add warmup support for check command

Follow-up from https://github.com/restic/restic/pull/5173
See also https://github.com/restic/restic/issues/3202

gilbsgilbs commented:
@MichaelEischer, sorry for the delay. Wrapping the backend was indeed way easier and I was able to write a test case.

pgreenland commented:
@MichaelEischer would be great to see this one merged. Have tested with OVHCloud's new cold storage tier, with a view to replacing my current custom backup tool with restic. Warmup during data checks appear to be working fine :-). Is there anything left to do on the PR that I may be able to help with?

gilbsgilbs commented:
@pgreenland Thanks for the bump. I just rebased my branch and resolved conflicts. Still active and willing to see this merged eventually :) .

gilbsgilbs commented:
Had to rebase again. I'm not sure how to get a maintainer's attention on this again, which is sad because they already reviewed the code and kindly gave me assistance on how to easily write the tests. I think the PR got so low in the PRs stack that nobody's even seeing the activity. I could send an email in, but there's such a backlog of PRs and issues that I don't want to add to the burden; I can already see why maintainers are overwhelmed and why this PR is not any more important than others.

I'll probably stop rebasing from now on. If a maintainer sees this, first of all thank you, and just ping me I'll do the rebase ASAP.

MichaelEischer commented on internal/checker/checker_test.go:
```diff
@@ -434,6 +437,61 @@ func TestCheckerModifiedData(t *testing.T) {
+	return nil
+}
+
+func TestCheckerWarmup(t *testing.T) {
```
From the looks of the test, there's no need for it to be in the `checker` package; move it to the `repository` package. There you can (by now) use the `runReadPacks` helper instead of `checkData` (and just drop reading snapshots).

MichaelEischer commented on changelog/unreleased/issue-3202:
```diff
@@ -0,0 +1,5 @@
+Enhancement: Add warmup support for check command on S3 backend
+
```
Please briefly mention the warmup support is still sort of experimental and behind a feature gate.

MichaelEischer reviewed (changes requested):
Sorry for the late review. I've indeed lost track of it and then didn't have much time to work on restic for a while. And I didn't notice that the PR is quite small and simple to review. I've started assigning myself to PRs to track which ones I want to focus one, which should at least help with quick re-reviews after the initial review.

Could you please rebase the PR again? It looks like I've managed to cause even more merge conflicts 🙈 . Besides that I only have two small comments.

gilbsgilbs commented:
@MichaelEischer 

> Sorry for the late review. I've indeed lost track of it and then didn't have much time to work on restic for a while. And I didn't notice that the PR is quite small and simple to review. I've started assigning myself to PRs to track which ones I want to focus one, which should at least help with quick re-reviews after the initial review.

No worries, really. If anything, I'm grateful that you came back to this PR eventually and that you keep maintaining this awesome project through the years.

I rebased against master and addressed your two comments. Please tell me if it looks good to you now or if you want me to address other things :) .

Thanks!

MichaelEischer reviewed (approved):
LGTM.
